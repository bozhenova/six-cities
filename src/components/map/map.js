import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { CITIES } from '../../../src/constants';
import leaflet from 'leaflet';

const MapConfig = {
  ID: `map`,
  ZOOM: 12,
  PIN_PATH: `img/pin.svg`,
  ACTIVE_PIN_PATH: `img/pin-active.svg`,
  PIN_SIZE: [30, 30],
  TITLE_LAYER: `https://{s}.basemaps.cartocdn.com/rastertiles/voyager/{z}/{x}/{y}{r}.png`
};

class Map extends Component {
  static propTypes = {
    coordinates: PropTypes.array.isRequired,
    currentCity: PropTypes.string.isRequired,
    selectedOffer: PropTypes.array
  };

  static defaultProps = {
    selectedOffer: []
  };

  group = null;
  icon = leaflet.icon({
    iconUrl: MapConfig.PIN_PATH,
    iconSize: MapConfig.PIN_SIZE
  });

  activeIcon = leaflet.icon({
    iconUrl: MapConfig.ACTIVE_PIN_PATH,
    iconSize: MapConfig.PIN_SIZE
  });

  addMarkers() {
    const { selectedOffer, coordinates } = this.props;
    const activeOffer = selectedOffer.length !== 0;
    this.group = leaflet.layerGroup().addTo(this.map);

    for (let item of coordinates) {
      leaflet.marker(item, { icon: this.icon }).addTo(this.group);
    }
    if (activeOffer) {
      leaflet
        .marker(selectedOffer, { icon: this.activeIcon })
        .addTo(this.group);
    }
  }

  initMap() {
    const city = CITIES.get(this.props.currentCity);

    this.map = leaflet.map(MapConfig.ID, {
      center: city,
      zoom: MapConfig.ZOOM,
      zoomControl: false,
      marker: true,
      scrollWheelZoom: false
    });
    this.map.setView(city, MapConfig.ZOOM);

    leaflet.tileLayer(MapConfig.TITLE_LAYER).addTo(this.map);
    this.addMarkers();
  }

  componentDidMount() {
    this.initMap();
  }

  componentDidUpdate() {
    const { currentCity, coordinates, selectedOffer } = this.props;

    this.group.clearLayers();
    this.map.setView(CITIES.get(currentCity), MapConfig.ZOOM);

    for (let item of coordinates) {
      leaflet.marker(item, { icon: this.icon }).addTo(this.group);
    }
    if (activeOffer) {
      leaflet
        .marker(selectedOffer, { icon: this.activeIcon })
        .addTo(this.group);
    }
  }

  render() {
    return (
      <div
        style={{ width: '100%', height: '100%', minHeight: '823px' }}
        className='cities__map'
        id='map'
      />
    );
  }
}

export default Map;
