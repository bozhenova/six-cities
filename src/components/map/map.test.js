import React from 'react';
import { Map } from './map';
import renderer from 'react-test-renderer';

const createMapBlock = () => {
  const div = global.document.createElement(`div`);
  div.setAttribute(`id`, `map`);
  global.document.body.appendChild(div);
};

const offers = [
  {
    city: {
      coords: [52.369553943508, 4.85309666406198],
      zoom: 12,
      name: 'Amsterdam'
    },
    place: {
      coords: [52.369553943508, 4.85309666406198],
      zoom: 12
    }
  }
];

const currentOfferCoords = [52.369553943508, 4.85309666406198];

describe('Map component renders correctly', () => {
  it('should render a map', () => {
    createMapBlock();
    const map = renderer
      .create(<Map offers={offers} currentOfferCoords={currentOfferCoords} />)
      .toJSON();
    expect(map).toMatchSnapshot();
  });
});
