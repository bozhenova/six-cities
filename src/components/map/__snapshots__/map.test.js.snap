// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map component renders correctly should render a map 1`] = `
<div
  className="cities__map"
  id="map"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
/>
`;
